test/functional/wallet_assumeutxo.py

```diff
@@ -32,19 +28,47 @@ def add_options(self, parser):
 
     def set_test_params(self):
         """Use the pregenerated, deterministic chain up to height 199."""
-        self.num_nodes = 2
+        self.num_nodes = 3
         self.rpc_timeout = 120
         self.extra_args = [
             [],
             [],
+            ["-prune=1"]
```

Comment:
You should use the `fastprune` option as well and check if there has already been some pruning happening when the test runs, i.e. check that there is some pruneheight. There isn't much of a difference if the option is set but nothing has actually been pruned yet.

Reply:
Thanks for the feedback. I added the `fastprune` option to my local copy and I'm working to understand how pruning operates.

When I tried loading the wallet during background synchronization ([here](https://github.com/bitcoin/bitcoin/pull/30455/files#diff-f763a7cf53e7a5183f66f2634baef84245919265a09d82b03dc435adea8b8faeR149)), I noticed that there wasn't a pruneheight set yet. So, I did a manual prune using `n2.pruneblockchain(START_HEIGHT)`. After pruning, I saw a pruneheight of 300, and the test continued as expected.

However, after the background validation and sync to the tip ([here](https://github.com/bitcoin/bitcoin/pull/30455/files#diff-f763a7cf53e7a5183f66f2634baef84245919265a09d82b03dc435adea8b8faeR200)), the pruneheight was reset to 0. When I tried to prune the blockchain again, the pruneheight remained at 0, no matter the value I passed to pruneblockchain. To further investigate, I mined an additional 500 blocks on n2 and manually pruned again with `n2.pruneblockchain(FINAL_HEIGHT)`. This time, the pruneheight was 299.

Despite this, the wallet restoration failed with this message:

> Wallet loading failed. Prune: last wallet synchronization goes beyond pruned data. You need to -reindex (download the whole blockchain again in case of pruned node)

I understand that a pruneheight of 299 means that the node keeps blocks from 299 onwards. Since the last processed block of the wallet is also 299 and the block hash matches, I expected the restoration to work. This issue also happens with lower pruneheight values, like 242.

Is this a bug, or am I missing something?

I can confirm that I am able to retrieve the block value with height `pruneheight`. When the pruneheight was 299, I validated that block 299 was still stored in the node. The same validation was done when the pruneheight was 242, confirming block 242 was also stored. So I am not sure why the wallet restoration fails

Reply:
> Thanks for the feedback. I added the fastprune option to my local copy and I'm working to understand how pruning operates.

In general, check `feature_index_prune.py`, there extensive tests using pruning and I think you can learn a bit more from that how much you have to mine for the fast pruning to work as you need it.

> However, after the background validation and sync to the tip ([here](https://github.com/bitcoin/bitcoin/pull/30455/files#diff-f763a7cf53e7a5183f66f2634baef84245919265a09d82b03dc435adea8b8faeR200)), the pruneheight was reset to 0. When I tried to prune the blockchain again, the pruneheight remained at 0, no matter the value I passed to pruneblockchain. To further investigate, I mined an additional 500 blocks on n2 and manually pruned again with n2.pruneblockchain(FINAL_HEIGHT). This time, the pruneheight was 299.

Pruning can feel weird when you operate with such low numbers, the answer is probably to play around with the number of mined blocks and the test mentioned above should help you with that. In general you have to understand that pruning means we delete db we no longer need (and fastprune makes them very small so we don't need them at a much faster rate). We no longer need them when nothing is in there that we need, i.e. no blocks that we might require are in there. Blocks are appended to these files as they come in so during normal IBD (in theory at least) we can always delete the db files in order they are created (FIFO) because the blocks come in order. But with assumeutxo the blocks don't come in order anymore. So that alone is weird and you need to keep that in mind. For the same reason `getblockfrompeer` can be hard to grasp, https://github.com/bitcoin/bitcoin/pull/23813 might be an interesting read about that. I think this is the reason you don't see results from `pruneblockchain` because there are no db files to be found that don't contain blocks we don't want to keep anymore. Mining more blocks in the test overall should fix this though, as you have seen already.

> Is this a bug, or am I missing something?

Hard to say for me with just what you have written. Can you push your current status to a branch somewhere and post the link, then I can take a look and play around with it.

Reply:
Thank you very much for your explanation and suggestions! I'll definitely check `feature_index_prune.py` and https://github.com/bitcoin/bitcoin/pull/23813 to better understand how this works.

Here is a branch where I pushed my current status: [debugging_assumeutxo_tests](https://github.com/alfonsoromanz/bitcoin/tree/debugging_assumeutxo_tests)

Reply:
Sorry for the long wait, I finally got around to look at your branch @alfonsoromanz . You are looking at this `wallet_last_processed_block` but that is coming out of node 0. When you load the wallet into node 2 it doesn't know anything about it, so this can be ignored. What you might be thinking of instead is the birthdate of the wallet, that's part of the backup afaict so that should be relevant. But the birthdate is earlier than 299 because the wallet seems to be created before a bunch of blocks are generated. So the error seems to make sense to me and you might want to create another wallet with a later birthdate which then should be successfully load at the place where you currently get the error with the older wallet.

Reply:
Ok, I see what you mean now. After chasing down some wrong leads I think I got it. The `rescan_height` is coming from the block locator stored on disk. This is a different value than `last_processed_block`. I found that there is not guaranteed that the latest block is written when a backup is created. Since the full database is included in the backup, there is indeed something there that would cause the wallet to only need to rescan after 299. So the correct behavior IMO would be that your case 2. shouldn't fail. I confirmed this with following change: #30678. I am suggesting there that this is changed but I am not sure this behavior is really so bad in practice that it will get merged. But at least I hope that answers your question.

Reply:
Thanks a lot @fjahr! Do you think that restoring a backup created in height `399` (scenario 3) would be enough for the purpose of this test?

Reply:
Yeah, I think that's enough for now. I am not convinced the other cases would cover additional interesting code paths but if I'm wrong the test can always be expanded later.

Reply:
Done. Thanks!